oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/config/group/EndpointNameGrouping.java

```diff
@@ -99,18 +105,30 @@ public Tuple2<String, Boolean> format(String serviceName, String endpointName) {
                     // Algorithm side should not return a pattern that has no {var} in it else this
                     // code may accidentally retrieve the size 1 queue created by unformatted endpoint
                     // The queue size is 10, which means only cache the first 10 formatted names.
-                    final ArrayBlockingQueue<String> formattedURIs = svrHttpUris.computeIfAbsent(
+                    final Queue<String> formattedURIs = svrHttpUris.computeIfAbsent(
                         formattedName._1(), k -> new ArrayBlockingQueue<>(10));
-                    if (formattedURIs.size() < 10) {
-                        // Try to push the raw URI as a candidate of formatted name.
-                        formattedURIs.offer(endpointName);
-                    }
+                    // Try to push the raw URI as a candidate of formatted name.
+                    formattedURIs.offer(endpointName);
                 } else {
                     svrHttpUris.computeIfAbsent(endpointName, k -> new ArrayBlockingQueue<>(1));
                 }
             }
         }
 
+        // If there are too many unformatted URIs, we will abandon the unformatted URIs to reduce
+        // the load of OAP and storage.
+        final var unformattedUrisOfService =
+            unformattedHttpUris.computeIfAbsent(serviceName, k -> ConcurrentHashMap.newKeySet());
+        if (!formattedName._2()) {
+            if (unformattedUrisOfService.size() < maxHttpUrisNumberPerService) {
+                unformattedUrisOfService.add(endpointName);
+            } else {
+                formattedName = new Tuple2<>(ABANDONED_ENDPOINT_NAME, true);
+            }
+        } else {
+            unformattedUrisOfService.remove(endpointName);
+        }
```

Comment:
@wu-sheng please review this part. The other part you commented on is unrelated.